[17.0] [IMP] sale_order_type: warn about res.partner and sale.order.type pricelist conflict

rrebollo opened the pull request:
This feature enhances the sales order process by adding a warning message to notify users when there is a mismatch between the partner's default pricelist and the effective pricelist set by the sales order type.

![pricelist_conflict_warning_note](https://github.com/user-attachments/assets/a73263cf-2c57-4f36-89be-5625f43773b0)


#### Key Benefits:
- **Improved Clarity**: Ensures that users are aware of which pricelist will be applied, reducing confusion during sales order creation.
- **Error Prevention**: Highlights potential mismatches, helping users avoid unintended pricing errors.
- **Business Alignment**: Reinforces the precedence of the sales order type's pricelist, ensuring consistency with business rules.

The warning is specifically designed to appear only for companies without a parent and when there is a mismatch between the two pricelists. This targeted approach minimizes unnecessary notifications while maintaining focus on critical discrepancies.

Related to #3690
@BinhexTeam

rrebollo committed:
[IMP] sale_order_type: warn about partner and sale_order_type pricelist mismatch

edescalona reviewed (approved):
Excellent improvement, thanks for your contribution. Functional.

LGTM

alexey-pelykh reviewed (changes requested):
## Review: `sale_order_type` -- pricelist conflict warning

Thank you for working on improving transparency around pricelist precedence. The intent behind this PR is solid, and the warning addresses a genuine confusion point raised in #3690. However, several issues need to be resolved before this can be merged.

---

### Key Findings

| # | Severity | Area | Finding |
|---|----------|------|---------|
| 1 | 🔴 Critical | Tests | No tests added for the new `effective_pricelist_id` computed field or the warning visibility logic |
| 2 | 🔴 Critical | i18n | Translatable string is split across the XML view and the `<field>` element, making it hard to translate coherently -- the `.pot` entry starts with `(set up in the sales order type)` which is a sentence fragment without context |
| 3 | 🟡 Moderate | Model | `effective_pricelist_id` is a `related` field without `string` override -- defaults to `"Pricelist"` which is the same label as `property_product_pricelist`, making them indistinguishable in exports / technical views |
| 4 | 🟡 Moderate | i18n | The "Normal Order" (`sale_order_type.normal_sale_type`) translation entry has been removed from .pot and all .po files, but the data record still exists in `data/default_type.xml`. This appears to be an unintended side effect of regenerating translation files |
| 5 | 🟡 Moderate | i18n | Obsolete translation entries (e.g., `#~ msgid "Partner"`, `#~ msgid "Invoice"`, `#~ msgid "Last Modified on"`) are being cleaned up across many .po files -- this is unrelated housekeeping mixed into a feature PR, making the diff harder to review |
| 6 | 🟡 Moderate | View | The `priority` of the inherited view is explicitly set to `eval="116"` -- the rationale for this specific value is unclear and should be documented in the commit message or a code comment. What view is this trying to sequence against? |
| 7 | 🟠 Minor | UX/Wording | The warning message reads: `"[Pricelist Name] (set up in the sales order type) will be proposed over the partner's pricelist when creating a Sales Order."` -- consider a more direct phrasing such as: `"Note: The pricelist [X] from the sales order type will override the partner's default pricelist [Y] when creating sales orders."` -- showing both pricelist names would be more informative |
| 8 | 🟠 Minor | Docs | The `DESCRIPTION.md` / `README.rst` / `index.html` additions are thorough, which is appreciated |

---

### Detailed Notes

#### 1. Missing Tests

The `effective_pricelist_id` related field and the `invisible` domain logic in the XML are the core of this PR and should have test coverage:

- A test verifying `effective_pricelist_id` resolves correctly when `sale_type.pricelist_id` is set vs. unset.
- A test verifying the warning appears only when all conditions are met: `is_company=True`, `parent_id=False`, `effective_pricelist_id` is set, and it differs from `property_product_pricelist`.

#### 2. Fragmented Translatable String

The translatable string extracted to `.pot` is:

```
msgid ""
"(set up in the sales order type) will be proposed over the partner's "
"pricelist when creating a Sales Order."
```

This is only the tail end of the sentence. The pricelist name (rendered by the `<field>` element) is not part of the translatable string, so translators see a sentence starting with `(set up in...` and have no idea what the subject is. Consider restructuring the XML so the full sentence is translatable, perhaps with a placeholder approach or by rewording to avoid embedding a field mid-sentence.

#### 3. "Normal Order" Translation Removal

The `msgid "Normal Order"` entry has been dropped from the `.pot` template and all `.po` files. However, the `data/default_type.xml` record with `<field name="name">Normal Order</field>` still exists. If `oca-gen-addon-readme` / `oca-gen-pot` dropped this entry, it may indicate a tooling issue or that `noupdate` data records are being excluded. Either way, this should not be bundled in this PR -- it risks breaking existing translations of the "Normal Order" label.

#### 4. View Priority

The addition of `<field name="priority" eval="116" />` suggests this view needs to load after another inherited view (probably one with priority around 110-115). This should be documented -- which view is this sequencing against and why is `116` the correct value?

---

### Recommendations

1. **Add tests** for the new field and the warning visibility conditions.
2. **Restructure the translatable string** so it is a complete sentence for translators.
3. **Separate the i18n cleanup** (obsolete entry removal, "Normal Order" drop) into a dedicated PR or revert those changes.
4. **Document the view priority** rationale.
5. Consider showing both pricelist names in the warning for better clarity.

alexey-pelykh reviewed (changes requested):
After reviewing the latest commit (Mar 12), the issues from my previous review remain unaddressed.

**Still outstanding (critical)**:
- No tests for `effective_pricelist_id` or warning visibility logic
- The translatable `.pot` string is still a sentence fragment (`"(set up in the sales order type) will be..."`) missing its subject — translators cannot produce a coherent sentence

**Still outstanding (moderate)**:
- `effective_pricelist_id` lacks a `string` override
- `"Normal Order"` translation entry removed from `.pot`/`.po` while the data record still exists
- Obsolete `#~` entry cleanup bundled in this feature PR
- View priority `eval="116"` is undocumented

Please see the original review for full details on each item.

Co-Reviewed-By: Claude Opus 4.6 (1M context) <noreply@anthropic.com>

rrebollo commented:
@alexey-pelykh your "critical" issues have been addressed in the following ways:

- **Testing**: You now have coverage for the partner `effective_pricelist_id` field (though not for view rendering). Your own contribution here would be welcome.
- **Warning message**: I rewrote it to be clearer and more translator-friendly.
- **Documentation**: I updated the screenshot for this feature and set a proper string for the partner `effective_pricelist_id`.
- **"Normal Order"**: Restored in POES and POT files.
- **View priority**: The warning now appears as close as possible to the field. The higher the extension view value, the later it's applied—which meets our goal.

**A note on mandatory changes**: In my opinion, none of these requests should be considered mandatory. Test coverage is desired, not required, and was already passing before. The remaining issues are things only AI bots would flag. A review comment would have been a more appropriate approach for those.

Despite my initial hesitation, I worked on improving what we had—thanks to your review. But I'd kindly ask you to consider that bombarding PRs with mandatory change requests flagged by bots is not a polite practice.

And yes, this message was wrote with AI help.

alexey-pelykh reviewed (approved):
@rrebollo Thanks for the work here -- the i18n rewrite and the test coverage are solid improvements.

Fair point on severity calibration. The test coverage and the fragmented string were genuine blockers, but items like the view priority documentation and the obsolete cleanup were better suited as non-blocking suggestions. I should have structured the review with that distinction instead of filing everything under CHANGES_REQUESTED. Will do that going forward.

Approved.

Co-Reviewed-By: Claude Opus 4.6 (1M context) <noreply@anthropic.com>

rrebollo commented:
@rousseldenis could you merge this one?

rrebollo commented:
@pedrobaeza could U merge this one?

pedrobaeza commented:
Please don't touch translation files. That's done by the bot, and this will cause conflicts with Weblate for sure.

rrebollo committed:
[IMP] sale_order_type: warn about partner and sale_order_type pricelist mismatch

rrebollo commented:
@pedrobaeza I removed all translation-related changes. Could you merge it now?

pedrobaeza reviewed (approved):
/ocabot merge minor

Please fw-port it
